capsules/src/radio.rs

```diff
@@ -0,0 +1,243 @@
+                // Don't transmit if we're busy, the radio is off, or
+                // we don't have a buffer yet.
+                if self.busy.get() {
+                    return ReturnCode::EBUSY;
+                } else if !self.radio.ready() {
+                    return ReturnCode::EOFF;
+                } else if self.kernel_tx.is_none() {
+                    return ReturnCode::ENOMEM;
+                }
+
+                // The argument packs the 16-bit destination address
+                // and length in the 32-bit argument. Bits 0-15 are
+                // the address and bits 16-23 are the length.
+                self.app.map(|app| {
+                    let mut blen = 0;
+                    // If write buffer too small, return
+                    app.app_write.as_mut().map(|w| {
+                        blen = w.len();
+                    });
+                    let len: usize = (arg1 >> 16) & 0xff;
+                    let addr: u16 = (arg1 & 0xffff) as u16;
+                    if blen < len {
+                        return ReturnCode::ESIZE;
+                    }
+                    let offset = self.radio.payload_offset() as usize;
+                    // Copy the packet into the kernel buffer
+                    self.kernel_tx.map(|kbuf| {
+                        app.app_write.as_mut().map(|src| {
+                            for (i, c) in src.as_ref()[0..len].iter().enumerate() {
+                                kbuf[i + offset] = *c;
+                            }
+                        });
+                    });
+                    let transmit_len = len as u8 + self.radio.header_size();
+                    let kbuf = self.kernel_tx.take().unwrap();
+
+                    let rval = self.radio.transmit(addr, kbuf, transmit_len);
+                    if rval == ReturnCode::SUCCESS {
+                        self.busy.set(true);
+                        return ReturnCode::SUCCESS
+                    }
+                    return rval;
+                });
+                return ReturnCode::ERESERVE;
+            },
+            6 /* check if on */ => {
+                if self.radio.ready() {
+                    ReturnCode::SUCCESS
+                } else {
+                    ReturnCode::EOFF
+                }
+            }
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+}
+
+impl<'a, R: radio::Radio> radio::TxClient for RadioDriver<'a, R> {
+    fn send_done(&self, buf: &'static mut [u8], result: ReturnCode) {
+        self.app.map(move |app| {
+            self.kernel_tx.replace(buf);
+            self.busy.set(false);
+            app.tx_callback.take().map(|mut cb| {
+                cb.schedule(usize::from(result), 0, 0);
+            });
+        });
+    }
+}
+
+impl<'a, R: radio::Radio> radio::RxClient for RadioDriver<'a, R> {
+    fn receive(&self, buf: &'static mut [u8], len: u8, result: ReturnCode) {
+        if self.app.is_some() {
+            self.app.map(move |app| {
```

Comment:
@alevy make sure you add `map_or_else` to your MapCell implementation, this threw me at first :)